### PR TITLE
Fix tsan failure in retry timer cancellation code.

### PR DESCRIPTION
**Pull request by markdroth**

### CODE REVIEW

File: src/core/ext/filters/client_channel/retry_filter.cc [hunk 1]
```diff
@@ -525,6 +528,58 @@ class RetryFilter::CallData {
...
+// those refs are not released until the LB call stack is destroyed.
+// When this object is destroyed, it will invoke the
+// on_call_stack_destruction closure from the surface.
+class RetryFilter::CallData::CallStackDestructionBarrier
+    : public RefCounted<CallStackDestructionBarrier, PolymorphicRefCount,
+                        kUnrefCallDtor> {
+ public:
+  CallStackDestructionBarrier() {}
+
+  ~CallStackDestructionBarrier() {
```
Comment (donnadionne) [inline src/core/ext/filters/client_channel/retry_filter.cc:553]: destructor may need an override.

Sanity had this error:
/root/.cache/bazel/_bazel_root/d2dc70c3d9da3fab488ba0dcbbd35051/execroot/com_github_grpc_grpc/src/core/ext/filters/client_channel/retry_filter.cc:553:3: error: annotate this function with 'override' or (rarely) 'final' [modernize-use-override,-warnings-as-errors]
  ~CallStackDestructionBarrier() {
  ^
                                 override

File: src/core/ext/filters/client_channel/retry_filter.cc [hunk 2]
```diff
@@ -525,6 +528,58 @@ class RetryFilter::CallData {
...
+    Ref().release();  // Ref held by callback.
+    grpc_closure* on_lb_call_destruction_complete =
+        calld->arena_->New<grpc_closure>();
+    GRPC_CLOSURE_INIT(on_lb_call_destruction_complete,
+                      OnLbCallDestructionComplete, this, nullptr);
+    return on_lb_call_destruction_complete;
+  }
+
+ private:
+  static void OnLbCallDestructionComplete(void* arg, grpc_error* error) {
```
Comment (donnadionne) [inline src/core/ext/filters/client_channel/retry_filter.cc:575]: error is unused leading to 
src/core/ext/filters/client_channel/retry_filter.cc:575:66: error: unused parameter 'error' [-Werror,-Wunused-parameter]
  static void OnLbCallDestructionComplete(void* arg, grpc_error* error) {

**Review by donnadionne — CHANGES_REQUESTED** [review]

File: src/core/ext/filters/client_channel/retry_filter.cc (hunk 1 above)
Comment (markdroth) [reply src/core/ext/filters/client_channel/retry_filter.cc]: Fixed.

File: src/core/ext/filters/client_channel/retry_filter.cc (hunk 2 above)
Comment (markdroth) [reply src/core/ext/filters/client_channel/retry_filter.cc]: Fixed.

**Review by donnadionne — APPROVED** [review]

**Comment by markdroth** [comment]
The "Basic Tests C# Windows" failure is an infrastructure timeout.

**Pull request merged**